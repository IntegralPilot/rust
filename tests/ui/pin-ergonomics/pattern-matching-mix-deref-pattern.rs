//@ revisions: normal pin_ergonomics deref_patterns both
//@ edition:2024
#![cfg_attr(any(pin_ergonomics, both), feature(pin_ergonomics))]
#![cfg_attr(any(deref_patterns, both), feature(deref_patterns))]
#![allow(incomplete_features)]

// This test verifies that the `pin_ergonomics` feature works well
// together with the `deref_patterns` feature under the error:
// "mix of deref patterns and normal constructors".

use std::pin::Pin;

struct Foo<T, U> {
    x: T,
    y: U,
}

struct Bar<T, U>(T, U);

enum Baz<T, U> {
    Foo(T, U),
    Bar { x: T, y: U },
}

fn foo_mut<T: Unpin, U: Unpin>(foo: Pin<&mut Foo<T, U>>) {
    match foo {
        Foo { x, y } => {}
        //[normal]~^ ERROR mismatched types
        //[deref_patterns]~^^ ERROR mix of deref patterns and normal constructors
        //[pin_ergonomics]~^^^ ERROR mix of deref patterns and normal constructors
        //[both]~^^^^ ERROR mix of deref patterns and normal constructors
        Pin { .. } => {}
    }
    let _ = || match foo {
        Foo { x, y } => {}
        //[normal]~^ ERROR mismatched types
        //[deref_patterns]~^^ ERROR mix of deref patterns and normal constructors
        //[pin_ergonomics]~^^^ ERROR mix of deref patterns and normal constructors
        //[both]~^^^^ ERROR mix of deref patterns and normal constructors
        Pin { .. } => {}
    };
}

fn foo_const<T: Unpin, U: Unpin>(foo: Pin<&Foo<T, U>>) {
    match foo {
        Foo { x, y } => {}
        //[normal]~^ ERROR mismatched types
        //[deref_patterns]~^^ ERROR mix of deref patterns and normal constructors
        //[pin_ergonomics]~^^^ ERROR mix of deref patterns and normal constructors
        //[both]~^^^^ ERROR mix of deref patterns and normal constructors
        Pin { .. } => {}
    }
    let _ = || match foo {
        Foo { x, y } => {}
        //[normal]~^ ERROR mismatched types
        //[deref_patterns]~^^ ERROR mix of deref patterns and normal constructors
        //[pin_ergonomics]~^^^ ERROR mix of deref patterns and normal constructors
        //[both]~^^^^ ERROR mix of deref patterns and normal constructors
        Pin { .. } => {}
    };
}

fn bar_mut<T: Unpin, U: Unpin>(bar: Pin<&mut Bar<T, U>>) {
    match bar {
        Bar(x, y) => {}
        //[normal]~^ ERROR mismatched types
        //[deref_patterns]~^^ ERROR mix of deref patterns and normal constructors
        //[pin_ergonomics]~^^^ ERROR mix of deref patterns and normal constructors
        //[both]~^^^^ ERROR mix of deref patterns and normal constructors
        Pin { .. } => {}
    }
    let _ = || match bar {
        Bar(x, y) => {}
        //[normal]~^ ERROR mismatched types
        //[deref_patterns]~^^ ERROR mix of deref patterns and normal constructors
        //[pin_ergonomics]~^^^ ERROR mix of deref patterns and normal constructors
        //[both]~^^^^ ERROR mix of deref patterns and normal constructors
        Pin { .. } => {}
    };
}

fn bar_const<T: Unpin, U: Unpin>(bar: Pin<&Bar<T, U>>) {
    match bar {
        Bar(x, y) => {}
        //[normal]~^ ERROR mismatched types
        //[deref_patterns]~^^ ERROR mix of deref patterns and normal constructors
        //[pin_ergonomics]~^^^ ERROR mix of deref patterns and normal constructors
        //[both]~^^^^ ERROR mix of deref patterns and normal constructors
        Pin { .. } => {}
    }
    let _ = || match bar {
        Bar(x, y) => {}
        //[normal]~^ ERROR mismatched types
        //[deref_patterns]~^^ ERROR mix of deref patterns and normal constructors
        //[pin_ergonomics]~^^^ ERROR mix of deref patterns and normal constructors
        //[both]~^^^^ ERROR mix of deref patterns and normal constructors
        Pin { .. } => {}
    };
}

fn main() {}
