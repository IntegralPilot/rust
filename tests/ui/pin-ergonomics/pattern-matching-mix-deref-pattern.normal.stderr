error[E0308]: mismatched types
  --> $DIR/pattern-matching-mix-deref-pattern.rs:27:9
   |
LL |     match foo {
   |           --- this expression has type `Pin<&mut Foo<T, U>>`
LL |         Foo { x, y } => {}
   |         ^^^^^^^^^^^^ expected `Pin<&mut Foo<T, U>>`, found `Foo<_, _>`
   |
   = note: expected struct `Pin<&mut Foo<T, U>>`
              found struct `Foo<_, _>`
help: consider dereferencing to access the inner value using the Deref trait
   |
LL |     match *foo {
   |           +

error[E0308]: mismatched types
  --> $DIR/pattern-matching-mix-deref-pattern.rs:35:9
   |
LL |     let _ = || match foo {
   |                      --- this expression has type `Pin<&mut Foo<T, U>>`
LL |         Foo { x, y } => {}
   |         ^^^^^^^^^^^^ expected `Pin<&mut Foo<T, U>>`, found `Foo<_, _>`
   |
   = note: expected struct `Pin<&mut Foo<T, U>>`
              found struct `Foo<_, _>`
help: consider dereferencing to access the inner value using the Deref trait
   |
LL |     let _ = || match *foo {
   |                      +

error[E0308]: mismatched types
  --> $DIR/pattern-matching-mix-deref-pattern.rs:46:9
   |
LL |     match foo {
   |           --- this expression has type `Pin<&Foo<T, U>>`
LL |         Foo { x, y } => {}
   |         ^^^^^^^^^^^^ expected `Pin<&Foo<T, U>>`, found `Foo<_, _>`
   |
   = note: expected struct `Pin<&Foo<T, U>>`
              found struct `Foo<_, _>`
help: consider dereferencing to access the inner value using the Deref trait
   |
LL |     match *foo {
   |           +

error[E0308]: mismatched types
  --> $DIR/pattern-matching-mix-deref-pattern.rs:54:9
   |
LL |     let _ = || match foo {
   |                      --- this expression has type `Pin<&Foo<T, U>>`
LL |         Foo { x, y } => {}
   |         ^^^^^^^^^^^^ expected `Pin<&Foo<T, U>>`, found `Foo<_, _>`
   |
   = note: expected struct `Pin<&Foo<T, U>>`
              found struct `Foo<_, _>`
help: consider dereferencing to access the inner value using the Deref trait
   |
LL |     let _ = || match *foo {
   |                      +

error[E0308]: mismatched types
  --> $DIR/pattern-matching-mix-deref-pattern.rs:65:9
   |
LL |     match bar {
   |           --- this expression has type `Pin<&mut Bar<T, U>>`
LL |         Bar(x, y) => {}
   |         ^^^^^^^^^ expected `Pin<&mut Bar<T, U>>`, found `Bar<_, _>`
   |
   = note: expected struct `Pin<&mut Bar<T, U>>`
              found struct `Bar<_, _>`
help: consider dereferencing to access the inner value using the Deref trait
   |
LL |     match *bar {
   |           +

error[E0308]: mismatched types
  --> $DIR/pattern-matching-mix-deref-pattern.rs:73:9
   |
LL |     let _ = || match bar {
   |                      --- this expression has type `Pin<&mut Bar<T, U>>`
LL |         Bar(x, y) => {}
   |         ^^^^^^^^^ expected `Pin<&mut Bar<T, U>>`, found `Bar<_, _>`
   |
   = note: expected struct `Pin<&mut Bar<T, U>>`
              found struct `Bar<_, _>`
help: consider dereferencing to access the inner value using the Deref trait
   |
LL |     let _ = || match *bar {
   |                      +

error[E0308]: mismatched types
  --> $DIR/pattern-matching-mix-deref-pattern.rs:84:9
   |
LL |     match bar {
   |           --- this expression has type `Pin<&Bar<T, U>>`
LL |         Bar(x, y) => {}
   |         ^^^^^^^^^ expected `Pin<&Bar<T, U>>`, found `Bar<_, _>`
   |
   = note: expected struct `Pin<&Bar<T, U>>`
              found struct `Bar<_, _>`
help: consider dereferencing to access the inner value using the Deref trait
   |
LL |     match *bar {
   |           +

error[E0308]: mismatched types
  --> $DIR/pattern-matching-mix-deref-pattern.rs:92:9
   |
LL |     let _ = || match bar {
   |                      --- this expression has type `Pin<&Bar<T, U>>`
LL |         Bar(x, y) => {}
   |         ^^^^^^^^^ expected `Pin<&Bar<T, U>>`, found `Bar<_, _>`
   |
   = note: expected struct `Pin<&Bar<T, U>>`
              found struct `Bar<_, _>`
help: consider dereferencing to access the inner value using the Deref trait
   |
LL |     let _ = || match *bar {
   |                      +

error: aborting due to 8 previous errors

For more information about this error, try `rustc --explain E0308`.
